File: golang-hedp/internal/blueprint/blueprint.go
// Package blueprint models the release graph: which Helm releases exist, what
// gates them on, what they depend on, and what they publish to their
// dependents.
//
// Dependencies here are between *releases*, not between charts. Helm's own
// subchart dependencies are resolved inside a single render; these edges span
// renders, which is what makes ordering (and therefore parallelism) a real
// question rather than a detail.
package blueprint

import (
	"fmt"
	"os"
	"sort"
	"strings"

	"sigs.k8s.io/yaml"
)

// Release is one node in the graph.
type Release struct {
	Name      string `json:"name"`
	Chart     string `json:"chart"`
	Namespace string `json:"namespace"`
	// Requires lists customer features that must all be enabled for this
	// release to be selected on its own. An empty list means "always on".
	Requires []string `json:"requires,omitempty"`
	// DependsOn lists releases that must render first. A dependency is pulled
	// into the plan even when its own Requires are not satisfied - that is the
	// entire point of an explicit edge.
	DependsOn []string `json:"dependsOn,omitempty"`
	// Values are the static per-release values, merged under the customer's.
	Values map[string]any `json:"values,omitempty"`
	// Exports are tiny templates evaluated after this release renders. The
	// results are handed to dependents under .Values.deps.<release>.
	Exports map[string]string `json:"exports,omitempty"`
}

// Blueprint is the whole graph, validated and topologically ordered.
type Blueprint struct {
	Releases []Release `json:"releases"`

	byName map[string]*Release
	// order is a global topological order; any subset of it is a valid order
	// for that subset.
	order map[string]int
	// dependents is the reverse index, used to explain impact.
	dependents map[string][]string
}

// Load reads and validates a blueprint file.
func Load(path string) (*Blueprint, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("read blueprint: %w", err)
	}
	return Parse(data)
}

// Parse validates a blueprint that is already in memory - the path taken when
// a versioned release set arrives as a bundle rather than a directory.
func Parse(data []byte) (*Blueprint, error) {
	var bp Blueprint
	if err := yaml.Unmarshal(data, &bp); err != nil {
		return nil, fmt.Errorf("parse blueprint: %w", err)
	}
	if err := bp.index(); err != nil {
		return nil, err
	}
	return &bp, nil
}

func (b *Blueprint) index() error {
	if len(b.Releases) == 0 {
		return fmt.Errorf("blueprint has no releases")
	}
	b.byName = make(map[string]*Release, len(b.Releases))
	b.dependents = make(map[string][]string, len(b.Releases))
	for i := range b.Releases {
		r := &b.Releases[i]
		if r.Name == "" {
			return fmt.Errorf("release %d has no name", i)
		}
		if r.Chart == "" {
			return fmt.Errorf("release %q has no chart", r.Name)
		}
		if _, dup := b.byName[r.Name]; dup {
			return fmt.Errorf("duplicate release %q", r.Name)
		}
		if r.Namespace == "" {
			r.Namespace = "default"
		}
		b.byName[r.Name] = r
	}
	for i := range b.Releases {
		r := &b.Releases[i]
		for _, dep := range r.DependsOn {
			if _, ok := b.byName[dep]; !ok {
				return fmt.Errorf("release %q depends on unknown release %q", r.Name, dep)
			}
			b.dependents[dep] = append(b.dependents[dep], r.Name)
		}
	}
	order, err := b.topoSort()
	if err != nil {
		return err
	}
	b.order = order
	return nil
}

// topoSort produces a deterministic global order and rejects cycles. A cycle
// in the release graph is a configuration bug that would otherwise surface as
// a deadlock or a missing export at render time, so it is caught at load.
func (b *Blueprint) topoSort() (map[string]int, error) {
	const (
		white = 0 // unvisited
		grey  = 1 // on the current path
		black = 2 // done
	)
	state := make(map[string]int, len(b.Releases))
	order := make(map[string]int, len(b.Releases))
	var out []string

	names := make([]string, 0, len(b.Releases))
	for i := range b.Releases {
		names = append(names, b.Releases[i].Name)
	}
	sort.Strings(names)

	var visit func(name string, path []string) error
	visit = func(name string, path []string) error {
		switch state[name] {
		case black:
			return nil
		case grey:
			return fmt.Errorf("dependency cycle: %s -> %s", strings.Join(path, " -> "), name)
		}
		state[name] = grey
		deps := append([]string(nil), b.byName[name].DependsOn...)
		sort.Strings(deps)
		for _, d := range deps {
			if err := visit(d, append(path, name)); err != nil {
				return err
			}
		}
		state[name] = black
		out = append(out, name)
		return nil
	}
	for _, n := range names {
		if err := visit(n, nil); err != nil {
			return nil, err
		}
	}
	for i, n := range out {
		order[n] = i
	}
	return order, nil
}

// Get returns a release by name.
func (b *Blueprint) Get(name string) (*Release, bool) {
	r, ok := b.byName[name]
	return r, ok
}

// Names returns all release names in topological order.
func (b *Blueprint) Names() []string {
	names := make([]string, 0, len(b.Releases))
	for i := range b.Releases {
		names = append(names, b.Releases[i].Name)
	}
	sort.Slice(names, func(i, j int) bool { return b.order[names[i]] < b.order[names[j]] })
	return names
}

// Features returns every feature named by any release's Requires. The service
// validates customer configs against this set, so a typo in a feature name is
// a 422 rather than a silently smaller render.
func (b *Blueprint) Features() []string {
	seen := map[string]bool{}
	var out []string
	for i := range b.Releases {
		for _, f := range b.Releases[i].Requires {
			if !seen[f] {
				seen[f] = true
				out = append(out, f)
			}
		}
	}
	sort.Strings(out)
	return out
}

// Plan is an ordered, wave-partitioned set of releases to render.
type Plan struct {
	// Waves partitions the selected releases so that every release in wave i
	// depends only on releases in waves < i. Everything inside a wave is
	// independent and can render in parallel.
	Waves [][]string
	// Selected is the flat set, in topological order.
	Selected []string
	// PulledIn records releases that were added purely to satisfy a
	// dependency, i.e. the customer's features did not select them directly.
	PulledIn []string
}

// Count returns the number of releases in the plan.
func (p Plan) Count() int { return len(p.Selected) }

// Width returns the widest wave, an upper bound on useful parallelism.
func (p Plan) Width() int {
	w := 0
	for _, wave := range p.Waves {
		if len(wave) > w {
			w = len(wave)
		}
	}
	return w
}

// PlanFor selects the releases a customer's feature set implies and partitions
// them into dependency waves.
//
// Selection is two-phase on purpose. First the feature gates pick the directly
// requested releases; then the dependency closure pulls in whatever those need.
// Reporting the second set separately (PulledIn) is what lets a customer see
// why they are paying for a release they never asked for.
func (b *Blueprint) PlanFor(features []string, all bool) Plan {
	enabled := make(map[string]bool, len(features))
	for _, f := range features {
		enabled[f] = true
	}

	direct := make(map[string]bool, len(b.Releases))
	for i := range b.Releases {
		r := &b.Releases[i]
		if all || b.satisfied(r, enabled) {
			direct[r.Name] = true
		}
	}

	selected := make(map[string]bool, len(direct))
	var closure func(name string)
	closure = func(name string) {
		if selected[name] {
			return
		}
		selected[name] = true
		for _, d := range b.byName[name].DependsOn {
			closure(d)
		}
	}
	for name := range direct {
		closure(name)
	}

	flat := make([]string, 0, len(selected))
	for name := range selected {
		flat = append(flat, name)
	}
	sort.Slice(flat, func(i, j int) bool { return b.order[flat[i]] < b.order[flat[j]] })

	var pulledIn []string
	for _, name := range flat {
		if !direct[name] {
			pulledIn = append(pulledIn, name)
		}
	}

	return Plan{Waves: b.waves(selected, flat), Selected: flat, PulledIn: pulledIn}
}

// waves assigns each release the depth of its longest dependency chain within
// the selection. Depth is the earliest wave a release can render in, so this
// is the schedule with the shortest critical path.
func (b *Blueprint) waves(selected map[string]bool, flat []string) [][]string {
	depth := make(map[string]int, len(flat))
	maxDepth := 0
	// flat is in topological order, so every dependency's depth is already known.
	for _, name := range flat {
		d := 0
		for _, dep := range b.byName[name].DependsOn {
			if !selected[dep] {
				continue
			}
			if dd := depth[dep] + 1; dd > d {
				d = dd
			}
		}
		depth[name] = d
		if d > maxDepth {
			maxDepth = d
		}
	}
	waves := make([][]string, maxDepth+1)
	for _, name := range flat {
		waves[depth[name]] = append(waves[depth[name]], name)
	}
	return waves
}

func (b *Blueprint) satisfied(r *Release, enabled map[string]bool) bool {
	for _, f := range r.Requires {
		if !enabled[f] {
			return false
		}
	}
	return true
}

// Validate checks that every release points at a chart that actually exists.
// Called once at startup against the catalog, so a missing chart is a boot
// failure rather than a per-request 500.
func (b *Blueprint) Validate(hasChart func(string) bool) error {
	var missing []string
	for i := range b.Releases {
		if !hasChart(b.Releases[i].Chart) {
			missing = append(missing, fmt.Sprintf("%s -> %s", b.Releases[i].Name, b.Releases[i].Chart))
		}
	}
	if len(missing) > 0 {
		return fmt.Errorf("blueprint references %d missing charts: %s", len(missing), strings.Join(missing, ", "))
	}
	return nil
}

File: rust-hedp/src/blueprint.rs
//! The release graph, ported from the Go service's internal/blueprint.
//!
//! Same semantics throughout: feature gates select releases, dependency edges
//! pull in whatever those need, cycles are rejected at load, and the selection
//! is partitioned into waves by longest dependency chain.

use std::collections::{BTreeMap, BTreeSet, HashMap};

use anyhow::{anyhow, Context, Result};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Release {
    pub name: String,
    pub chart: String,
    #[serde(default)]
    pub namespace: String,
    #[serde(default)]
    pub requires: Vec<String>,
    #[serde(default, rename = "dependsOn")]
    pub depends_on: Vec<String>,
    #[serde(default)]
    pub values: serde_json::Value,
    #[serde(default)]
    pub exports: BTreeMap<String, String>,
}

#[derive(Debug, Deserialize)]
struct BlueprintFile {
    releases: Vec<Release>,
}

pub struct Blueprint {
    pub releases: Vec<Release>,
    by_name: HashMap<String, usize>,
    /// order is a global topological order; any subset of it is valid for that
    /// subset.
    order: HashMap<String, usize>,
}

#[derive(Debug, Default, Serialize)]
pub struct Plan {
    pub waves: Vec<Vec<String>>,
    pub selected: Vec<String>,
    pub pulled_in: Vec<String>,
}

impl Plan {
    pub fn count(&self) -> usize {
        self.selected.len()
    }
    pub fn width(&self) -> usize {
        self.waves.iter().map(|w| w.len()).max().unwrap_or(0)
    }
}

impl Blueprint {
    pub fn load(path: &std::path::Path) -> Result<Self> {
        let data = std::fs::read_to_string(path)
            .with_context(|| format!("read blueprint {}", path.display()))?;
        Self::parse(&data)
    }

    pub fn parse(data: &str) -> Result<Self> {
        let file: BlueprintFile = serde_yaml_ng::from_str(data).context("parse blueprint")?;
        if file.releases.is_empty() {
            return Err(anyhow!("blueprint has no releases"));
        }

        let mut releases = file.releases;
        let mut by_name = HashMap::with_capacity(releases.len());
        for (i, r) in releases.iter_mut().enumerate() {
            if r.name.is_empty() {
                return Err(anyhow!("release {i} has no name"));
            }
            if r.chart.is_empty() {
                return Err(anyhow!("release {} has no chart", r.name));
            }
            if r.namespace.is_empty() {
                r.namespace = "default".into();
            }
            if by_name.insert(r.name.clone(), i).is_some() {
                return Err(anyhow!("duplicate release {}", r.name));
            }
        }
        for r in &releases {
            for dep in &r.depends_on {
                if !by_name.contains_key(dep) {
                    return Err(anyhow!("release {} depends on unknown release {dep}", r.name));
                }
            }
        }

        let order = topo_sort(&releases, &by_name)?;
        Ok(Blueprint {
            releases,
            by_name,
            order,
        })
    }

    pub fn get(&self, name: &str) -> Option<&Release> {
        self.by_name.get(name).map(|i| &self.releases[*i])
    }

    pub fn names(&self) -> Vec<String> {
        let mut n: Vec<String> = self.releases.iter().map(|r| r.name.clone()).collect();
        n.sort_by_key(|name| self.order[name]);
        n
    }

    pub fn features(&self) -> Vec<String> {
        let set: BTreeSet<&String> = self.releases.iter().flat_map(|r| r.requires.iter()).collect();
        set.into_iter().cloned().collect()
    }

    pub fn validate(&self, has_chart: impl Fn(&str) -> bool) -> Result<()> {
        let missing: Vec<String> = self
            .releases
            .iter()
            .filter(|r| !has_chart(&r.chart))
            .map(|r| format!("{} -> {}", r.name, r.chart))
            .collect();
        if !missing.is_empty() {
            return Err(anyhow!(
                "blueprint references {} missing charts: {}",
                missing.len(),
                missing.join(", ")
            ));
        }
        Ok(())
    }

    /// plan_for selects what a customer's features imply and partitions it into
    /// dependency waves.
    pub fn plan_for(&self, features: &[String], all: bool) -> Plan {
        let enabled: BTreeSet<&str> = features.iter().map(|s| s.as_str()).collect();

        let mut direct = BTreeSet::new();
        for r in &self.releases {
            if all || r.requires.iter().all(|f| enabled.contains(f.as_str())) {
                direct.insert(r.name.clone());
            }
        }

        let mut selected: BTreeSet<String> = BTreeSet::new();
        let mut stack: Vec<String> = direct.iter().cloned().collect();
        while let Some(name) = stack.pop() {
            if !selected.insert(name.clone()) {
                continue;
            }
            if let Some(r) = self.get(&name) {
                for d in &r.depends_on {
                    if !selected.contains(d) {
                        stack.push(d.clone());
                    }
                }
            }
        }

        let mut flat: Vec<String> = selected.iter().cloned().collect();
        flat.sort_by_key(|n| self.order[n]);

        let pulled_in: Vec<String> = flat
            .iter()
            .filter(|n| !direct.contains(*n))
            .cloned()
            .collect();

        // Depth is the longest dependency chain inside the selection, i.e. the
        // earliest wave a release can run in.
        let mut depth: HashMap<&str, usize> = HashMap::with_capacity(flat.len());
        let mut max_depth = 0;
        for name in &flat {
            let mut d = 0;
            if let Some(r) = self.get(name) {
                for dep in &r.depends_on {
                    if selected.contains(dep) {
                        d = d.max(depth.get(dep.as_str()).copied().unwrap_or(0) + 1);
                    }
                }
            }
            depth.insert(name.as_str(), d);
            max_depth = max_depth.max(d);
        }
        let mut waves = vec![Vec::new(); max_depth + 1];
        for name in &flat {
            waves[depth[name.as_str()]].push(name.clone());
        }

        Plan {
            waves,
            selected: flat,
            pulled_in,
        }
    }
}

fn topo_sort(
    releases: &[Release],
    by_name: &HashMap<String, usize>,
) -> Result<HashMap<String, usize>> {
    const WHITE: u8 = 0;
    const GREY: u8 = 1;
    const BLACK: u8 = 2;

    let mut state: HashMap<&str, u8> = HashMap::with_capacity(releases.len());
    let mut out: Vec<&str> = Vec::with_capacity(releases.len());

    let mut names: Vec<&str> = releases.iter().map(|r| r.name.as_str()).collect();
    names.sort();

    // Iterative DFS: a deep chain should not be able to blow the stack.
    for root in names {
        if state.get(root).copied().unwrap_or(WHITE) == BLACK {
            continue;
        }
        let mut stack: Vec<(&str, usize)> = vec![(root, 0)];
        while let Some((name, idx)) = stack.pop() {
            if idx == 0 {
                match state.get(name).copied().unwrap_or(WHITE) {
                    BLACK => continue,
                    GREY => {
                        let path: Vec<&str> = stack.iter().map(|(n, _)| *n).collect();
                        return Err(anyhow!("dependency cycle: {} -> {name}", path.join(" -> ")));
                    }
                    _ => {}
                }
                state.insert(name, GREY);
            }

            let deps = {
                let r = &releases[by_name[name]];
                let mut d: Vec<&str> = r.depends_on.iter().map(|s| s.as_str()).collect();
                d.sort();
                d
            };

            if idx < deps.len() {
                stack.push((name, idx + 1));
                let next = deps[idx];
                if state.get(next).copied().unwrap_or(WHITE) == GREY {
                    return Err(anyhow!("dependency cycle: {name} -> {next}"));
                }
                stack.push((next, 0));
            } else {
                state.insert(name, BLACK);
                out.push(name);
            }
        }
    }

    Ok(out
        .into_iter()
        .enumerate()
        .map(|(i, n)| (n.to_string(), i))
        .collect())
}
